File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:

//same the combination 1 
    //create ans and ds to store the answer and elements
    //call find (0 , target , candidate , ans ,ds );
    //return ans;

//in find (index , target , arr , ans ,ds)
    //if target ==0 ans.push(ds);

    //loop from index -> arr.size()
        //check the arr[i] == arr[i-1] then continue
        //check arr[i] > target  ====> break;
    
    //ds.push(arr[i])
    //call find (index+1 , target - arr[i] , arr , ans, ds)
    //ds.pop();

void find(int index ,int target , vector<int>&arr ,vector<vector<int>>&ans , vector<int>& ds){
    if(target == 0) {
        ans.push_back(ds);
        return ;
    }
    for(int i=index ;i<arr.size() ;i++){
        if(i > index && arr[i]== arr[i-1]) continue;
        if(arr[i] > target)break;

        ds.push_back(arr[i]);
        find(i +1 ,target -arr[i] ,arr ,ans , ds);
        ds.pop_back();
    }
}
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>>ans;
        vector<int>ds;
        find(0,target ,candidates ,ans ,ds);
        return ans;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res = set()
        nums.sort()

        def dfs(i , curr , total):
            if total == target :
                res.add(tuple(curr))
                return 

            if i >= len(nums) or total > target :
                return 

            curr.append(nums[i])
            dfs(i+1 , curr , total + nums[i])
            curr.pop()

            while i+1 < len(nums) and nums[i] == nums[i+1]:
                i+=1

            dfs(i+1 , curr , total)

        dfs(0,[],0)
        
        return [list(i) for i in res]

        